<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Said el Bouhaddani" />

<meta name="date" content="2020-06-16" />

<title>ML course - Part 1a - PCA</title>
<!-- Material Design fonts -->
<link rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700">
<link rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/icon?family=Material+Icons">
<script src="ML_part1_Rpractical_PCA_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="ML_part1_Rpractical_PCA_files/bootstrap-3.3.7/css/bootstrap.min.css" rel="stylesheet" />
<script src="ML_part1_Rpractical_PCA_files/bootstrap-3.3.7/js/bootstrap.min.js"></script>
<script src="ML_part1_Rpractical_PCA_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<script src="ML_part1_Rpractical_PCA_files/navigation-1.1/tabsets.js"></script>
<script src="ML_part1_Rpractical_PCA_files/navigation-1.1/codefolding.js"></script>
<link href="ML_part1_Rpractical_PCA_files/magnific-popup-1.1.0/magnific-popup.css" rel="stylesheet" />
<script src="ML_part1_Rpractical_PCA_files/magnific-popup-1.1.0/jquery.magnific-popup.min.js"></script>
<link href="ML_part1_Rpractical_PCA_files/bootstrap_material-0.1/bootstrap-material-design.min.css" rel="stylesheet" />
<link href="ML_part1_Rpractical_PCA_files/bootstrap_material-0.1/ripples.min.css" rel="stylesheet" />
<script src="ML_part1_Rpractical_PCA_files/bootstrap_material-0.1/material.min.js"></script>
<script src="ML_part1_Rpractical_PCA_files/bootstrap_material-0.1/ripples.min.js"></script>
<link href="ML_part1_Rpractical_PCA_files/material-0.1/material.css" rel="stylesheet" />
<script src="ML_part1_Rpractical_PCA_files/material-0.1/material.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    background-color: #ffffff;
    color: #a0a0a0;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #a0a0a0;  padding-left: 4px; }
div.sourceCode
  { color: #1f1c1b; background-color: #ffffff; }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span. { color: #1f1c1b; } /* Normal */
code span.al { color: #bf0303; background-color: #f7e6e6; font-weight: bold; } /* Alert */
code span.an { color: #ca60ca; } /* Annotation */
code span.at { color: #0057ae; } /* Attribute */
code span.bn { color: #b08000; } /* BaseN */
code span.bu { color: #644a9b; font-weight: bold; } /* BuiltIn */
code span.cf { color: #1f1c1b; font-weight: bold; } /* ControlFlow */
code span.ch { color: #924c9d; } /* Char */
code span.cn { color: #aa5500; } /* Constant */
code span.co { color: #898887; } /* Comment */
code span.cv { color: #0095ff; } /* CommentVar */
code span.do { color: #607880; } /* Documentation */
code span.dt { color: #0057ae; } /* DataType */
code span.dv { color: #b08000; } /* DecVal */
code span.er { color: #bf0303; text-decoration: underline; } /* Error */
code span.ex { color: #0095ff; font-weight: bold; } /* Extension */
code span.fl { color: #b08000; } /* Float */
code span.fu { color: #644a9b; } /* Function */
code span.im { color: #ff5500; } /* Import */
code span.in { color: #b08000; } /* Information */
code span.kw { color: #1f1c1b; font-weight: bold; } /* Keyword */
code span.op { color: #1f1c1b; } /* Operator */
code span.ot { color: #006e28; } /* Other */
code span.pp { color: #006e28; } /* Preprocessor */
code span.re { color: #0057ae; background-color: #e0e9f8; } /* RegionMarker */
code span.sc { color: #3daee9; } /* SpecialChar */
code span.ss { color: #ff5500; } /* SpecialString */
code span.st { color: #bf0303; } /* String */
code span.va { color: #0057ae; } /* Variable */
code span.vs { color: #bf0303; } /* VerbatimString */
code span.wa { color: #bf0303; } /* Warning */
</style>


<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("toc");
});
</script>

<!-- code folding -->

</head>

<body>

<div class="header-panel shadow z-2">
    <div class="container-fluid">
        <div class="row">
            <div class="col-xs-3">
        <div id="header">
    <h1 class="title">ML course - Part 1a - PCA</h1>
                <h4 class="author">Said el Bouhaddani</h4>
                <h4 class="date">2020-06-16</h4>
        </div>
    </div>
</div>
</div>
</div>


<div class="container-fluid main-container">
    <div class="row">
      <nav class="col-xs-3 menu">
        <div id="toc">
        <ul>
        <li><a href="#introduction">Introduction</a></li>
        <li><a href="#load-real-data">Load real data</a></li>
        <li><a href="#analysis-with-principal-components-analysis">Analysis with Principal Components Analysis</a></li>
        <li><a href="#references">References</a></li>
        </ul>
        </div>
        
        
        
      </nav>
     <div class="pages col-xs-9">
     <div class="row">
       <div class="col-xs-10">



<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>This is a file with exercises for the ML 2020 course. The exercises are divided into two parts. This is part 1a: Principal components analysis.</p>
<div id="pca-recap" class="section level2">
<h2>PCA recap</h2>
<p>As discussed in the lecture, the PCA solution for finding <span class="math inline">\(w\)</span> that maximizes the variance of the projected data matrix <span class="math inline">\(Xw\)</span> is the first right singular vector of <span class="math inline">\(X\)</span>. Further directions of maximal variance are given by the subsequent singular vectors. We typically collect all of these vectors in a matrix <span class="math inline">\(W\)</span>. The scores are given by <span class="math inline">\(T=XW\)</span>, one per column. Based on these two matrices, we can approximate <span class="math inline">\(X\)</span> by <span class="math inline">\(TW^\top\)</span> (which is nothing more than the projection of <span class="math inline">\(X\)</span> onto <span class="math inline">\(WW^\top\)</span>).</p>
</div>
<div id="quick-tips" class="section level2">
<h2>Quick tips</h2>
<ul>
<li>To obtain the right singular vectors of a matrix <span class="math inline">\(A\)</span>, one can use <code>svd(A,nu=0)</code>. The output is a list with elements <code>d</code> and <code>v</code>. The <code>v</code> contain the singular vectors. These elements can be accessed with the <code>$</code> key.</li>
<li>Matrices can be subsetted with the <code>[</code> operator, e.g. <code>A$vectors[,1]</code>.</li>
<li>The scores can be obtained by multiplying <span class="math inline">\(X\)</span> and <span class="math inline">\(W\)</span>, e.g. <code>T = X %*% W</code>.</li>
</ul>
</div>
<div id="exercises" class="section level2">
<h2>Exercises</h2>
<ul>
<li>Normal exercises
<ul>
<li>Show that <span class="math inline">\(TW^\top\)</span> is “nothing more than” <span class="math inline">\(XWW^\top\)</span></li>
<li>Simulate two vectors <code>x1 = rnorm(100)</code> and <code>x2 = rnorm(100)</code>. Apply PCA and inspect the weights of the data matrix <code>X = cbind(x1, x2)</code>. Recall, the weights are given by <code>svd(X,0)$v</code>.</li>
<li>Now introduce correlation, for example by running <code>x2 = x1 + rnorm(100, sd = 0.1)</code>. Now run the PCA decomposition again and inspect the weights.
<ul>
<li>How does correlation affect the weights?</li>
</ul></li>
</ul></li>
<li>Advanced exercises
<ul>
<li>What is wrong with the following optimization problem: <span class="math inline">\(\max_w w^\top X^\top X w\)</span> (with <span class="math inline">\(w\)</span> unrestricted)? What is the largest possible covariance we can obtain?</li>
<li>Formulate the optimization problem for finding the second direction <span class="math inline">\(w_2\)</span> given the first direction <span class="math inline">\(w_1\)</span>. Hint, <span class="math inline">\(w_2\)</span> should be orthogonal to <span class="math inline">\(w_1\)</span>.</li>
</ul></li>
</ul>
<div id="solutions" class="section level3">
<h3>Solutions</h3>
<ul>
<li>Normal exercises
<ul>
<li>Note that by definition, <span class="math inline">\(T = XW\)</span>, therefore <span class="math inline">\(TW^\top = XWW^\top\)</span>.</li>
<li>The weights should be random for each run, as it totally depends on the variances of the columns.</li>
<li>If the variables are correlated, the weights are much more stable. They should be around <span class="math inline">\(\sqrt{0.5}\)</span>.</li>
</ul></li>
<li>More difficult exercises, you can safely skip it for now
<ul>
<li>Let <span class="math inline">\(w\)</span> go to infinity, then the covariance gets larger and larger. Hence, the optimization problem does not admit a solution.</li>
<li>It is <span class="math inline">\(\max_w w^\top X^\top X w\)</span> s.t. <span class="math inline">\(w^\top w = 1\)</span> and <span class="math inline">\(w^\top w_1 = 0\)</span></li>
</ul></li>
</ul>
</div>
</div>
</div>
<div id="load-real-data" class="section level1">
<h1>Load real data</h1>
<p>We have transcriptomic and metabolomic measurements from a Finnish population cohort, as part of the DILGOM study. The transcriptomic measurements can be found at ArrayExpress (<a href="http://www.ebi.ac.uk/arrayexpress/" class="uri">http://www.ebi.ac.uk/arrayexpress/</a>) under accession number <code>E-TABM-1036</code> (<code>E-TABM-1036.processed.1.zip</code>). The metabolite measurements are attached as supplemental material at <span class="citation">(Inouye et al. 2010)</span> (<code>msb201093-sup-0002.zip</code>). These data are already prepared and on Github, so download the file <code>rna_metab.RData</code>.</p>
<div id="load-the-data" class="section level2">
<h2>Load the data</h2>
<p>This code chunk loads the transcriptomic data into memory.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">load</span>(<span class="st">&quot;rna_metab.RData&quot;</span>)</span></code></pre></div>
<p>There are two datasets in your environment: <code>rna</code> (transcripts) and <code>metab</code> (metabolites). Check with <code>ls()</code>.</p>
</div>
<div id="inspect-the-data-descriptives" class="section level2">
<h2>Inspect the data: descriptives</h2>
<p><strong>Packages needed</strong></p>
<ul>
<li><code>install.packages("gplots")</code></li>
</ul>
<p>A heatmap of metabolites is plotted. There are some groups of correlated variables it seems.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>gplots<span class="op">::</span><span class="kw">heatmap.2</span>(<span class="kw">cor</span>(metab,<span class="dt">use =</span> <span class="st">&#39;pair&#39;</span>), <span class="dt">dendrogram=</span><span class="st">&#39;none&#39;</span>, <span class="dt">Rowv=</span>F, <span class="dt">Colv=</span>F,<span class="dt">trace=</span><span class="st">&#39;n&#39;</span>,</span>
<span id="cb2-2"><a href="#cb2-2"></a>                  <span class="dt">breaks=</span><span class="kw">seq</span>(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>,<span class="dt">length.out =</span> <span class="dv">25</span>), <span class="dt">col=</span>gplots<span class="op">::</span>bluered)</span></code></pre></div>
<p><img src="Figs/Heatmap%20of%20correlations-1.png" width="8cm" height="6cm" /></p>
<p>Boxplots provide a good summary to compare the distribution of the variables relative to each other. Properties such as comparable means, variances and symmetry are often good to have. For sake of visualisation, we only consider the first 100 transcripts.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="kw">boxplot</span>(rna[,<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>])</span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="kw">boxplot</span>(metab)</span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code></pre></div>
<p><img src="Figs/Boxplots-1.png" width="8cm" height="6cm" /> The distributions are quite symmetric and the scale is comparable across variables in each data set.</p>
</div>
</div>
<div id="analysis-with-principal-components-analysis" class="section level1">
<h1>Analysis with Principal Components Analysis</h1>
<div id="run-pca-on-the-data" class="section level2">
<h2>Run PCA on the data</h2>
<p>We perform PCA by calculating the SVD. Note that this is much faster than running <code>eigen</code> on the matrix <span class="math inline">\(X^\top X\)</span>, which is 7385 times 7385.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>pca.rna =<span class="st"> </span><span class="kw">svd</span>(rna, <span class="dv">0</span>, <span class="dv">2</span>)</span>
<span id="cb4-2"><a href="#cb4-2"></a>pca.metab &lt;-<span class="st"> </span><span class="kw">svd</span>(metab, <span class="dv">0</span>, <span class="dv">2</span>)</span>
<span id="cb4-3"><a href="#cb4-3"></a></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb4-5"><a href="#cb4-5"></a><span class="kw">plot</span>(rna <span class="op">%*%</span><span class="st"> </span>pca.rna<span class="op">$</span>v, <span class="dt">main =</span> <span class="st">&quot;RNA PCA plot of the scores&quot;</span>,<span class="dt">xlab=</span><span class="ot">NA</span>,<span class="dt">ylab=</span><span class="ot">NA</span>)</span>
<span id="cb4-6"><a href="#cb4-6"></a><span class="kw">plot</span>(metab <span class="op">%*%</span><span class="st"> </span>pca.metab<span class="op">$</span>v, <span class="dt">main =</span> <span class="st">&quot;Metabolites PCA plot of the scores&quot;</span>,<span class="dt">xlab=</span><span class="ot">NA</span>,<span class="dt">ylab=</span><span class="ot">NA</span>)</span></code></pre></div>
<p><img src="Figs/PCA%20fit-1.png" width="3600" /></p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code></pre></div>
<p>No particular structure or outliers are visible.</p>
<p>Now to plot the weights for the two directions. Note that the RNA dataset has around 7000 weights to plot. For the metabolite weights, we add a coloring based on the type of metabolite.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="kw">library</span>(magrittr)</span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="kw">library</span>(ggplot2)</span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="kw">library</span>(gridExtra)</span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="kw">library</span>(OmicsPLS)</span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="kw">library</span>(illuminaHumanv3.db) <span class="co"># BiocManager::install(&quot;illuminaHumanv3.db&quot;)</span></span>
<span id="cb6-6"><a href="#cb6-6"></a><span class="co"># Color names</span></span>
<span id="cb6-7"><a href="#cb6-7"></a>LLmodule &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;ILMN_1690209&quot;</span>,<span class="st">&#39;ILMN_1766551&#39;</span>, <span class="st">&#39;ILMN_1749131&#39;</span>, <span class="st">&#39;ILMN_1688423&#39;</span>, </span>
<span id="cb6-8"><a href="#cb6-8"></a>              <span class="st">&#39;ILMN_2102670&#39;</span>, <span class="st">&#39;ILMN_1792323&#39;</span>, <span class="st">&#39;ILMN_1899034&#39;</span>, <span class="st">&#39;ILMN_1806721&#39;</span>, </span>
<span id="cb6-9"><a href="#cb6-9"></a>              <span class="st">&#39;ILMN_1695530&#39;</span>, <span class="st">&#39;ILMN_1726114&#39;</span>, <span class="st">&#39;ILMN_1751625&#39;</span>, <span class="st">&#39;ILMN_1726114&#39;</span>, </span>
<span id="cb6-10"><a href="#cb6-10"></a>              <span class="st">&#39;ILMN_1753648&#39;</span>, <span class="st">&#39;ILMN_1779043&#39;</span>)</span>
<span id="cb6-11"><a href="#cb6-11"></a>LLnr &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">colnames</span>(rna) <span class="op">%in%</span><span class="st"> </span>LLmodule)</span>
<span id="cb6-12"><a href="#cb6-12"></a>rna_genenames &lt;-<span class="st"> </span><span class="kw">select</span>(illuminaHumanv3.db, </span>
<span id="cb6-13"><a href="#cb6-13"></a>                        <span class="dt">keys =</span> <span class="kw">colnames</span>(rna)[LLnr], </span>
<span id="cb6-14"><a href="#cb6-14"></a>                        <span class="dt">keytype =</span> <span class="st">&quot;PROBEID&quot;</span>, <span class="dt">columns =</span> <span class="st">&quot;SYMBOL&quot;</span>)[,<span class="dv">2</span>]</span>
<span id="cb6-15"><a href="#cb6-15"></a></span>
<span id="cb6-16"><a href="#cb6-16"></a>name_col &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">+</span><span class="st"> </span><span class="kw">sapply</span>( <span class="co">#First sapply loops over column names</span></span>
<span id="cb6-17"><a href="#cb6-17"></a>  <span class="dt">X =</span> <span class="kw">colnames</span>(metab),</span>
<span id="cb6-18"><a href="#cb6-18"></a>  <span class="dt">FUN =</span> <span class="cf">function</span>(arg){</span>
<span id="cb6-19"><a href="#cb6-19"></a>    <span class="kw">crossprod</span>(</span>
<span id="cb6-20"><a href="#cb6-20"></a>      <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>), <span class="co"># Weights to be used as categories</span></span>
<span id="cb6-21"><a href="#cb6-21"></a>      <span class="kw">sapply</span>(<span class="kw">c</span>(<span class="st">&quot;VLDL&quot;</span>, <span class="st">&quot;LDL&quot;</span>, <span class="st">&quot;IDL&quot;</span>, <span class="st">&quot;HDL&quot;</span>,<span class="st">&quot;FA&quot;</span>), <span class="co"># metabolite classes</span></span>
<span id="cb6-22"><a href="#cb6-22"></a>             <span class="cf">function</span>(arg2){<span class="kw">grepl</span>(arg2, arg)} <span class="co"># compare class of metabolites</span></span>
<span id="cb6-23"><a href="#cb6-23"></a>      )</span>
<span id="cb6-24"><a href="#cb6-24"></a>    )</span>
<span id="cb6-25"><a href="#cb6-25"></a>    }</span>
<span id="cb6-26"><a href="#cb6-26"></a>  )</span>
<span id="cb6-27"><a href="#cb6-27"></a>name_col &lt;-<span class="st"> </span><span class="kw">factor</span>(name_col, </span>
<span id="cb6-28"><a href="#cb6-28"></a>                   <span class="dt">levels =</span> <span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">2</span>,<span class="dv">4</span><span class="op">:</span><span class="dv">6</span>,<span class="dv">1</span>), </span>
<span id="cb6-29"><a href="#cb6-29"></a>                   <span class="dt">labels =</span> <span class="kw">c</span>(<span class="st">&quot;VLDL&quot;</span>, <span class="st">&quot;LDL&quot;</span>, <span class="st">&quot;IDL&quot;</span>, <span class="st">&quot;HDL&quot;</span>,<span class="st">&quot;FA&quot;</span>,<span class="st">&quot;Other&quot;</span>))</span>
<span id="cb6-30"><a href="#cb6-30"></a></span>
<span id="cb6-31"><a href="#cb6-31"></a><span class="co"># alpmetab &lt;- loadings(fit, &quot;Yjoint&quot;, 1:2) %&gt;%  # Retreive loadings</span></span>
<span id="cb6-32"><a href="#cb6-32"></a><span class="co">#   abs %&gt;% # Absolute loading values for positive weights</span></span>
<span id="cb6-33"><a href="#cb6-33"></a><span class="co">#   rowSums %&gt;% # Sum over the components</span></span>
<span id="cb6-34"><a href="#cb6-34"></a><span class="co">#   sqrt + (name_col!=&quot;Other&quot;) # Take square root</span></span>
<span id="cb6-35"><a href="#cb6-35"></a></span>
<span id="cb6-36"><a href="#cb6-36"></a><span class="co">######### Plot loadings with ggplot </span><span class="al">###</span></span>
<span id="cb6-37"><a href="#cb6-37"></a>p_metab &lt;-<span class="st"> </span><span class="kw">ggplot</span>(<span class="kw">data.frame</span>(<span class="dt">x =</span> pca.metab<span class="op">$</span>v[,<span class="dv">1</span>], <span class="dt">y =</span> pca.metab<span class="op">$</span>v[, <span class="dv">2</span>]), <span class="kw">aes</span>(<span class="dt">x =</span> x, <span class="dt">y =</span> y)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb6-38"><a href="#cb6-38"></a><span class="co">##################### Add all layers </span><span class="al">###</span></span>
<span id="cb6-39"><a href="#cb6-39"></a><span class="st">  </span><span class="kw">theme_bw</span>() <span class="op">+</span></span>
<span id="cb6-40"><a href="#cb6-40"></a><span class="st">  </span><span class="kw">coord_fixed</span>(<span class="dt">ratio =</span> <span class="dv">1</span>, <span class="dt">xlim=</span><span class="kw">c</span>(<span class="op">-</span>.<span class="dv">2</span>,.<span class="dv">2</span>),<span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span>.<span class="dv">2</span>,.<span class="dv">2</span>)) <span class="op">+</span></span>
<span id="cb6-41"><a href="#cb6-41"></a><span class="st">  </span><span class="kw">geom_point</span>( <span class="co"># Set color and size</span></span>
<span id="cb6-42"><a href="#cb6-42"></a>    <span class="kw">aes</span>(<span class="dt">col=</span>name_col, <span class="dt">size =</span> <span class="kw">I</span>(<span class="dv">1</span><span class="op">+</span>(name_col<span class="op">%in%</span><span class="kw">c</span>(<span class="st">&quot;VLDL&quot;</span>,<span class="st">&quot;HDL&quot;</span>))), </span>
<span id="cb6-43"><a href="#cb6-43"></a>          <span class="dt">shape =</span> name_col),<span class="dt">show.legend =</span> T) <span class="op">+</span></span>
<span id="cb6-44"><a href="#cb6-44"></a><span class="st">  </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="st">&quot;right&quot;</span>) <span class="op">+</span></span>
<span id="cb6-45"><a href="#cb6-45"></a><span class="st">  </span><span class="kw">scale_color_discrete</span>(<span class="dt">name=</span><span class="st">&quot;Metabolite</span><span class="ch">\n</span><span class="st">Group&quot;</span>,</span>
<span id="cb6-46"><a href="#cb6-46"></a>                       <span class="dt">labels=</span><span class="kw">c</span>(<span class="st">&quot;VLDL&quot;</span>, <span class="st">&quot;LDL&quot;</span>, <span class="st">&quot;IDL&quot;</span>, <span class="st">&quot;HDL&quot;</span>,<span class="st">&quot;FA&quot;</span>,<span class="st">&quot;Other&quot;</span>)) <span class="op">+</span></span>
<span id="cb6-47"><a href="#cb6-47"></a><span class="st">  </span><span class="kw">guides</span>(<span class="dt">size=</span>F) <span class="op">+</span><span class="st"> </span><span class="kw">scale_shape_discrete</span>(<span class="dt">name=</span><span class="st">&quot;Metabolite</span><span class="ch">\n</span><span class="st">Group&quot;</span>,</span>
<span id="cb6-48"><a href="#cb6-48"></a>                                <span class="dt">labels=</span><span class="kw">c</span>(<span class="st">&quot;VLDL&quot;</span>, <span class="st">&quot;LDL&quot;</span>, <span class="st">&quot;IDL&quot;</span>, <span class="st">&quot;HDL&quot;</span>,<span class="st">&quot;FA&quot;</span>,<span class="st">&quot;Other&quot;</span>)) <span class="op">+</span></span>
<span id="cb6-49"><a href="#cb6-49"></a><span class="st">  </span><span class="kw">scale_shape_manual</span>(<span class="dt">name=</span><span class="st">&quot;Metabolite</span><span class="ch">\n</span><span class="st">Group&quot;</span>, <span class="dt">values=</span><span class="kw">c</span>(<span class="dv">15</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">17</span>,<span class="dv">5</span>,<span class="dv">6</span>)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb6-50"><a href="#cb6-50"></a><span class="st">  </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&quot;Metabolite joint loadings&quot;</span>,</span>
<span id="cb6-51"><a href="#cb6-51"></a>       <span class="dt">x =</span> <span class="st">&quot;First Joint Loadings&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Second Joint Loadings&quot;</span>) <span class="op">+</span></span>
<span id="cb6-52"><a href="#cb6-52"></a><span class="st">  </span><span class="kw">theme</span>(<span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">face=</span><span class="st">&#39;bold&#39;</span>),</span>
<span id="cb6-53"><a href="#cb6-53"></a>        <span class="dt">legend.title=</span><span class="kw">element_text</span>(<span class="dt">face=</span><span class="st">&#39;bold&#39;</span>)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb6-54"><a href="#cb6-54"></a><span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="dv">0</span>) <span class="op">+</span><span class="st"> </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="dv">0</span>)</span>
<span id="cb6-55"><a href="#cb6-55"></a></span>
<span id="cb6-56"><a href="#cb6-56"></a>alprna &lt;-<span class="st"> </span>pca.rna<span class="op">$</span>v <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">raise_to_power</span>(<span class="dv">2</span>) <span class="op">%&gt;%</span><span class="st"> </span>rowSums</span>
<span id="cb6-57"><a href="#cb6-57"></a>alprna[<span class="op">-</span>(<span class="kw">order</span>(alprna,<span class="dt">decreasing=</span>T)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>])] =<span class="st"> </span><span class="dv">0</span></span>
<span id="cb6-58"><a href="#cb6-58"></a>alprna &lt;-<span class="st"> </span><span class="kw">sign</span>(alprna)</span>
<span id="cb6-59"><a href="#cb6-59"></a>toprna &lt;-<span class="st"> </span><span class="kw">which</span>(alprna<span class="op">&gt;</span><span class="dv">0</span>)</span>
<span id="cb6-60"><a href="#cb6-60"></a>names_rna &lt;-<span class="st"> </span><span class="kw">mapIds</span>(illuminaHumanv3.db, </span>
<span id="cb6-61"><a href="#cb6-61"></a>       <span class="dt">keys =</span> <span class="kw">colnames</span>(rna)[toprna], </span>
<span id="cb6-62"><a href="#cb6-62"></a>       <span class="dt">keytype =</span> <span class="st">&quot;PROBEID&quot;</span>, </span>
<span id="cb6-63"><a href="#cb6-63"></a>       <span class="dt">column =</span> <span class="st">&quot;SYMBOL&quot;</span>,</span>
<span id="cb6-64"><a href="#cb6-64"></a>       <span class="dt">multiVals =</span> <span class="st">&#39;first&#39;</span>)</span>
<span id="cb6-65"><a href="#cb6-65"></a>names_rna[<span class="kw">which</span>(<span class="kw">is.na</span>(names_rna))] &lt;-<span class="st"> &quot;?&quot;</span></span>
<span id="cb6-66"><a href="#cb6-66"></a><span class="co">######### Plot loadings with OmicsPLS plot method </span><span class="al">###</span></span>
<span id="cb6-67"><a href="#cb6-67"></a>p_rna &lt;-<span class="st"> </span><span class="kw">ggplot</span>(<span class="kw">data.frame</span>(<span class="dt">x =</span> pca.rna<span class="op">$</span>v[, <span class="dv">1</span>], <span class="dt">y =</span> pca.rna<span class="op">$</span>v[, <span class="dv">2</span>]), </span>
<span id="cb6-68"><a href="#cb6-68"></a>                <span class="kw">aes</span>(<span class="dt">x =</span> x, <span class="dt">y =</span> y),</span>
<span id="cb6-69"><a href="#cb6-69"></a>                <span class="dt">alpha =</span> alprna,</span>
<span id="cb6-70"><a href="#cb6-70"></a>                <span class="kw">aes</span>(<span class="dt">label =</span> <span class="ot">NA</span>)) <span class="op">+</span></span>
<span id="cb6-71"><a href="#cb6-71"></a><span class="st">    </span><span class="co">##################### Add all layers </span><span class="al">###</span></span>
<span id="cb6-72"><a href="#cb6-72"></a><span class="st">  </span><span class="kw">theme_bw</span>() <span class="op">+</span></span>
<span id="cb6-73"><a href="#cb6-73"></a><span class="st">  </span><span class="kw">coord_fixed</span>(.<span class="dv">8</span>, <span class="kw">c</span>(<span class="op">-</span>.<span class="dv">15</span>,.<span class="dv">15</span>),<span class="kw">c</span>(<span class="op">-</span>.<span class="dv">15</span>,.<span class="dv">15</span>)) <span class="op">+</span></span>
<span id="cb6-74"><a href="#cb6-74"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">alpha =</span> <span class="fl">0.5</span>, <span class="dt">col =</span> <span class="st">&#39;grey&#39;</span>) <span class="op">+</span></span>
<span id="cb6-75"><a href="#cb6-75"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">data =</span> <span class="kw">data.frame</span>(<span class="dt">x =</span> pca.rna<span class="op">$</span>v[LLnr, <span class="dv">1</span>], <span class="dt">y =</span> pca.rna<span class="op">$</span>v[LLnr, <span class="dv">2</span>]),</span>
<span id="cb6-76"><a href="#cb6-76"></a>             <span class="dt">shape =</span> <span class="dv">2</span>, <span class="dt">col =</span> <span class="dv">2</span>, <span class="dt">size =</span> <span class="dv">2</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb6-77"><a href="#cb6-77"></a><span class="st">  </span><span class="kw">geom_text</span>(<span class="dt">data =</span> <span class="kw">data.frame</span>(<span class="dt">x=</span>pca.rna<span class="op">$</span>v[toprna,<span class="dv">1</span>],<span class="dt">y=</span>pca.rna<span class="op">$</span>v[toprna,<span class="dv">2</span>]),</span>
<span id="cb6-78"><a href="#cb6-78"></a>            <span class="dt">hjust =</span> <span class="kw">rep</span>(<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">0</span>), <span class="dt">length.out =</span> <span class="kw">length</span>(toprna)),</span>
<span id="cb6-79"><a href="#cb6-79"></a>            <span class="kw">aes</span>(<span class="dt">label =</span> names_rna)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb6-80"><a href="#cb6-80"></a><span class="st">  </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&quot;Transcript joint loadings&quot;</span>,</span>
<span id="cb6-81"><a href="#cb6-81"></a>       <span class="dt">x =</span> <span class="st">&quot;First Joint Loadings&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Second Joint Loadings&quot;</span>) <span class="op">+</span></span>
<span id="cb6-82"><a href="#cb6-82"></a><span class="st">  </span><span class="kw">theme</span>(<span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">face=</span><span class="st">&#39;bold&#39;</span>)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb6-83"><a href="#cb6-83"></a><span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="dv">0</span>) <span class="op">+</span><span class="st"> </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="dv">0</span>)</span>
<span id="cb6-84"><a href="#cb6-84"></a></span>
<span id="cb6-85"><a href="#cb6-85"></a><span class="co">## Finally plot both plots in one figure.</span></span>
<span id="cb6-86"><a href="#cb6-86"></a><span class="kw">grid.arrange</span>(p_metab, p_rna, <span class="dt">ncol=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="Figs/plot%20loadings-1.png" width="3600" /></p>
</div>
<div id="exercises-1" class="section level2">
<h2>Exercises</h2>
<ul>
<li>Why is it better to derive singular vectors from <span class="math inline">\(X\)</span> than eigenvectors from <span class="math inline">\(X^\top X\)</span>, where <span class="math inline">\(X\)</span> is the <code>rna</code> dataset?</li>
<li>Identify the most extreme expression probe with <code>identify</code>. To this end, first run <code>plot(pca.rna$v)</code>.
<ul>
<li>To which gene does it map? Tip: use google.</li>
<li>Why does it have the largest weight?</li>
</ul></li>
<li>What kind of clusters appear in the metabolite weights?</li>
<li>From previous analysis, it was shown that the LL module seemed to be connected to lipid metabolism. The genes in the LL module are shown as red triangles. Are they among the top genes?</li>
</ul>
<div id="solutions-1" class="section level3">
<h3>Solutions</h3>
<ul>
<li>Here, <code>t(rna) %*% rna</code> is a 7385 by 7385 matrix. This does probably not fit into memory. Furthermore, with <code>eigen</code> you cannot limit the number of PC’s to be calculated.</li>
<li>Run <code>identify(pca.rna$v, labels = colnames(rna))</code> and click on the most extreme values. Then click ESC. It is the second gene in the dataset. When inspecting the boxplots of the first 10 genes, it appears that this gene has a higher variance.</li>
<li>Roughly speaking: VLDL and HDL.</li>
<li>No. Because PCA does not look at the relation between <code>rna</code> and <code>metab</code>.</li>
</ul>
</div>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1>References</h1>
<div id="refs" class="references">
<div id="ref-Inouye2010msb">
<p>Inouye, Michael, Johannes Kettunen, Pasi Soininen, Kaisa Silander, Samuli Ripatti, Linda S Kumpula, Eija Hämäläinen, et al. 2010. “Metabonomic, Transcriptomic, and Genomic Variation of a Population Cohort.” <em>Molecular Systems Biology</em> 6 (1). <a href="https://doi.org/10.1038/msb.2010.93">https://doi.org/10.1038/msb.2010.93</a>.</p>
</div>
</div>
</div>

    </div>
    <div class="col-xs-2">
        </div>
  </div>
  </div>
  </div>
  <div class="row">
    </div>
  </div>

<script>
$(document).ready(function () {
  // add bootstrap table styles to pandoc tables
  $('tr.header').parent('thead').parent('table').addClass('table table-striped table-hover');

    var images = $('.pages img');
  images.filter(function() {
      if ($(this).parent().attr("class") == "figure") {
          return(false)
      } else {
          return(true);
      }
  }).wrap("<div class='figure'></div>");
  images.addClass("image-thumb").wrap("<div class='panel-body'></div>");
  $('.figure p.caption').wrap("<div class='panel-footer'></div>");
  $('.figure').addClass('panel panel-default');
  
    $('.pages img')
 	  .addClass("image-lb");
  $('.pages').magnificPopup({
	      type:'image',
	      closeOnContentClick: false,
	      closeBtnInside: false,
        delegate: 'img',
	      gallery: {enabled: false },
          removalDelay: 500,
          callbacks: {
              beforeOpen: function() {
                // just a hack that adds mfp-anim class to markup
                this.st.image.markup = this.st.image.markup.replace('mfp-figure', 'mfp-figure mfp-with-anim');
              }
          },
          mainClass: 'mfp-move-from-top',
	      image: {
	        verticalFit: true,
            titleSrc: 'alt'
	      }
 	    });
 	
    
    $('#toc ul li').first().addClass("active");
    $('#toc ul li').attr("data-target", function() {
        return($(this).children("a").attr("href"));
    })
    $('body .section.level1').first().addClass("active");
    $('body .section.level1').not('.active').hide();
    
    $('#toc a[href*="#"]').click(function() {

      var id = $(this).attr("href");
      if (id === "#") return;
      if (id.substring(0, 8) === "#dyntab-") return;
      toggle_page(id);

      // Menu
      var menu_entry = $(".menu li[data-target='"+id+"']");
      menu_entry.addClass("active");
      $(".menu li").not(menu_entry).removeClass("active"); 
      

    });

    function toggle_page(id) {
      $(".page").not(page).removeClass("active").hide();
      window.page = id;
      var page = $(window.page);
      window.location.hash = window.page;
      //$(this).addClass("active");

      page.show();

      var totop = setInterval(function () {
        $(".pages").animate({scrollTop: 0}, 0);
      }, 10);

      setTimeout(function () {
        page.addClass("active");
        setTimeout(function () {
          clearInterval(totop);
        }, 1000);
      }, 100);

      window.dispatchEvent(new Event('resize'));

    }


    $(".menu li").click(function () {

      toggle_page($(this).data("target"));

      // Menu
      if (!$(this).data("target")) return;
      if ($(this).is(".active")) return;
      $(".menu li").not($(this)).removeClass("active");
      $(this).addClass("active");

    });
  
    


    window.page = window.location.hash;
    if (window.page != "") {
      $(".menu").find("li[data-target=" + window.page + "]").trigger("click");
    }

    /* init material bootstrap js */
    $.material.init();
});
</script>




<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
